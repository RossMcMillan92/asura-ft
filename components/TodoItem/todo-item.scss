@import '../../styles/settings/breakpoints.scss';
@import '../../styles/settings/pallete.scss';

$todo-item__size: 30px;
$todo-item__spacing: 15px;
$todo-item__spacing--tablet: 25px;

.todo-item {
  position: relative;
  display: flex;
}

.todo-item__label {
  display: block;
  font-size: 16px;
  line-height: 1.4;
  word-break: break-word;
  display: flex;
  flex-grow: 1;
  align-items: center;
  color: $base-text-color;
  padding: $todo-item__spacing ($todo-item__spacing * 4) $todo-item__spacing
    $todo-item__spacing;
  border: solid 1px rgba(0, 0, 0, 0);
  border-width: 1px 0;

  &:focus-within,
  &:hover,
  &:active,
  .todo-item:focus-within & {
    background-color: #fafafa;
  }

  @media (min-width: $breakpoint-tablet) {
    font-size: 19px;
    padding: $todo-item__spacing--tablet ($todo-item__spacing--tablet * 4)
      $todo-item__spacing--tablet $todo-item__spacing--tablet;
  }
}

.todo-item__input-wrapper {
  position: relative;
  margin-right: $todo-item__spacing;

  @media (min-width: $breakpoint-tablet) {
    margin-right: $todo-item__spacing--tablet;
  }
}

.todo-item__delete-button-container {
  position: absolute;
  right: $todo-item__spacing;
  top: 50%;
  transform: translateY(-50%);

  @media (min-width: $breakpoint-tablet) {
    right: $todo-item__spacing--tablet;
  }
}

.todo-item__delete-button-icon {
  display: block;

  .todo-item:hover & {
    opacity: 1;
  }
  @media (min-width: $breakpoint-tablet) {
    opacity: 0;

    .todo-item:hover &,
    .todo-item:focus-within & {
      opacity: 1;
    }
  }

  // Ensure the button is visible on larger touch devices
  @media (hover: none) {
    opacity: 1;
  }
}
