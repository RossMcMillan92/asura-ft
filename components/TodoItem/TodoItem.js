import './todo-item.scss'

import Button from '../Button/Button'
import CheckboxInput from '../CheckboxInput/CheckboxInput'
import PropTypes from 'prop-types'
import React from 'react'
import classnames from 'classnames'

const TodoItem = ({ checked, id, label, onDelete, name, onChange, value }) => (
  <li className={classnames('todo-item', { 'is-checked': checked })}>
    <label className={'todo-item__label'} htmlFor={id}>
      <div className={'todo-item__input-wrapper'}>
        <CheckboxInput
          id={id}
          type="checkbox"
          name={name}
          value={value}
          checked={!!checked}
          onChange={onChange}
        />
      </div>
      {label}
    </label>

    <div className="todo-item__delete-button-container">
      <Button
        aria-label={`Delete item: ${label}.`}
        id={`${id}-delete-button`}
        data-testid="delete-button"
        onClick={onDelete}
        theme="bare"
      >
        <DeleteButtonIcon />
      </Button>
    </div>
  </li>
)

const DeleteButtonIcon = () => (
  <svg
    className="todo-item__delete-button-icon"
    viewBox="0 0 1024 1024"
    width="14"
    height="14"
    role="presentation"
  >
    <path
      fill="currentColor"
      d="M1014.662 822.66c-0.004-0.004-0.008-0.008-0.012-0.010l-310.644-310.65 310.644-310.65c0.004-0.004 0.008-0.006 0.012-0.010 3.344-3.346 5.762-7.254 7.312-11.416 4.246-11.376 1.824-24.682-7.324-33.83l-146.746-146.746c-9.148-9.146-22.45-11.566-33.828-7.32-4.16 1.55-8.070 3.968-11.418 7.31 0 0.004-0.004 0.006-0.008 0.010l-310.648 310.652-310.648-310.65c-0.004-0.004-0.006-0.006-0.010-0.010-3.346-3.342-7.254-5.76-11.414-7.31-11.38-4.248-24.682-1.826-33.83 7.32l-146.748 146.748c-9.148 9.148-11.568 22.452-7.322 33.828 1.552 4.16 3.97 8.072 7.312 11.416 0.004 0.002 0.006 0.006 0.010 0.010l310.65 310.648-310.65 310.652c-0.002 0.004-0.006 0.006-0.008 0.010-3.342 3.346-5.76 7.254-7.314 11.414-4.248 11.376-1.826 24.682 7.322 33.83l146.748 146.746c9.15 9.148 22.452 11.568 33.83 7.322 4.16-1.552 8.070-3.97 11.416-7.312 0.002-0.004 0.006-0.006 0.010-0.010l310.648-310.65 310.648 310.65c0.004 0.002 0.008 0.006 0.012 0.008 3.348 3.344 7.254 5.762 11.414 7.314 11.378 4.246 24.684 1.826 33.828-7.322l146.746-146.748c9.148-9.148 11.57-22.454 7.324-33.83-1.552-4.16-3.97-8.068-7.314-11.414z"
      stroke="inherit"
    />
  </svg>
)

TodoItem.propTypes = {
  checked: PropTypes.bool.isRequired,
  onDelete: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired
}

export default TodoItem
