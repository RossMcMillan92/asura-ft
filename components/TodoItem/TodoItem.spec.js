import { mount, shallow } from 'enzyme'

import React from 'react'
import TodoItem from './TodoItem'
import { act } from 'react-dom/test-utils'

const props = {
  checked: false,
  id: 'checkbox-id',
  name: 'checkbox-name',
  onChange: jest.fn(),
  onDelete: jest.fn(),
  value: 'value',
  label: 'Todo item label'
}
const getWrapper = additionalProps =>
  mount(<TodoItem {...props} {...additionalProps} />)
const getShallowWrapper = additionalProps =>
  shallow(<TodoItem {...props} {...additionalProps} />)

describe('TodoItem', () => {
  it('matches the snapshot', () => {
    const wrapper = getShallowWrapper()
    expect(wrapper).toMatchSnapshot()
  })

  describe('when the rendered checkbox is changed', () => {
    it('calls the passed onChange prop', () => {
      const wrapper = getWrapper()
      act(() => {
        wrapper.find('input').simulate('change')
      })
      expect(props.onChange).toBeCalled()
    })
  })

  describe('when the delete button is clicked', () => {
    it('calls the passed onDelete prop', () => {
      const wrapper = getWrapper()
      act(() => {
        wrapper
          .find('#checkbox-id-delete-button')
          .hostNodes()
          .simulate('click')
      })
      expect(props.onDelete).toBeCalled()
    })
  })

  describe('when the prop checked is passed as true', () => {
    const checked = true
    const wrapper = getWrapper({ checked })

    it('sets the input to checked', () => {
      expect(wrapper.find('input').prop('checked')).toBe(checked)
    })

    it('appends the class to the className', () => {
      expect(wrapper.find('.todo-item').prop('className')).toContain(
        'todo-item is-checked'
      )
    })
  })
})
