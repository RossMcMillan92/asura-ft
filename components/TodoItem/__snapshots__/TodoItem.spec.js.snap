// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem matches the snapshot 1`] = `
<li
  className="todo-item"
>
  <label
    className="todo-item__label"
    htmlFor="checkbox-id"
  >
    <div
      className="todo-item__input-wrapper"
    >
      <CheckboxInput
        checked={false}
        id="checkbox-id"
        name="checkbox-name"
        onChange={[MockFunction]}
        type="checkbox"
        value="value"
      />
    </div>
    Todo item label
  </label>
  <div
    className="todo-item__delete-button-container"
  >
    <Button
      aria-label="Delete item: Todo item label."
      data-testid="delete-button"
      id="checkbox-id-delete-button"
      onClick={[MockFunction]}
      theme="bare"
    >
      <DeleteButtonIcon />
    </Button>
  </div>
</li>
`;
